### PR TITLE
FLAG 78 : Add unit test for patient flags

### PR DESCRIPTION
**Pull request by ManojLL**
## Description of what I changed

added unit tests for flag service class
## Issue I worked on

see https://openmrs.atlassian.net/browse/FLAG-78

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.
  
  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java [hunk 1]
```diff
@@ -1,17 +1,22 @@
 package org.openmrs.module.patientflags;
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 
 import junit.framework.Assert;
```
Comment (wikumChamith) [inline api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:7]: Can we switch to org.junit.Assert?

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java [hunk 2]
```diff
@@ -20,30 +25,76 @@
...
 	
 	/**
 	 * Tests of the generateFlagsForPatient(Patient patient) method
 	 */
 	@Test
 	public void generateFlagsForPatient_shouldAcceptNullParameter() throws Exception {
-		Context.getService(FlagService.class).generateFlagsForPatient(new Patient(), null);
+		Patient patient = Context.getService(PatientService.class).getPatient(2);
+		List<Flag> flags = flagService.generateFlagsForPatient(patient, null);
+		Assert.assertTrue(flags.size() > 0);
```
Comment (wikumChamith) [inline api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:53]: Can we use static imports here?

Comment (wikumChamith) [inline api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:53]: we could use !flags.isEmpty() here

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java [hunk 3]
```diff
@@ -20,30 +25,76 @@
...
 	public void getFlaggedPatients_shouldAcceptNullFlagParameter() throws Exception {
-		Context.getService(FlagService.class).getFlaggedPatients(new Flag(), null);
+		Flag flag = flagService.getFlag(1);
+		Cohort cohort = flagService.getFlaggedPatients(flag, null);
+		Assert.assertFalse(cohort.isEmpty());
+	}
+
+	@Test
+	public void getFlaggedPatients_shouldReturnListOfFlags() throws Exception {
+		HashMap<Object, Object> context = new HashMap<Object, Object>();
```
Comment (wikumChamith) [inline api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:76]: Map<Object, Object> context = new HashMap<>();

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java (hunk 1 above)
Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java]: I switch to static imports

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java (hunk 2 above)
Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java]: I updated the changes

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java [hunk 4]
```diff
@@ -1,17 +1,25 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
 
+import org.junit.Before;
 import org.junit.Test;
+import org.openmrs.Cohort;
 import org.openmrs.Patient;
+import org.openmrs.api.PatientService;
 import org.openmrs.api.context.Context;
 import org.openmrs.module.patientflags.Flag;
```
Comment (wikumChamith) [inline api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:18]: These are shown as unused imports, do we need these?

Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:18]: They will use when we writing test for tags and priority services.

Comment (wikumChamith) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:18]: There's no need to keep imports for future use. If they aren't currently being used, you could remove them.

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java [hunk 5]
```diff
@@ -20,30 +28,76 @@
...
+		executeDataSet(TEST_DATASET_FILE);
+		flagService = Context.getService(FlagService.class);
+		authenticate();
+	}
 	
 	/**
 	 * Tests of the generateFlagsForPatient(Patient patient) method
 	 */
 	@Test
 	public void generateFlagsForPatient_shouldAcceptNullParameter() throws Exception {
```
Comment (wikumChamith) [inline api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:53]: Do these tests throw any exceptions?

Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:53]: these are not throwing any exception

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java [hunk 6]
```diff
@@ -20,30 +28,76 @@
  */
 public class FlagServiceTest extends BaseModuleContextSensitiveTest {
 
-	// TODO: create some actual methods that test the flagging of patients?
+	protected static final String XML_DATASET_PATH = "org/openmrs/module/patientflags/include/";
```
[the remark this answers is not in the document]
Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:31]: we can have one venerable for this, but other modules also use two variables to import the test datset

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java (hunk 5 above)
Comment (wikumChamith) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:53]: So can we remove the `throws Exception ` ?

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wikumChamith) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java:31]: ok let's keep this

**Comment by wikumChamith** [comment]
Can you explain to me again why we are using two dataset files?

**Comment by ManojLL** [comment]
> Can you explain to me again why we are using two dataset files?

FlagTest class contains test to validate flags logic [(validate_shouldRejectInvalidGroovyCriteria())](https://github.com/openmrs/openmrs-module-patientflags/blob/e969c9795a191200ef1614d74a4cd43dfc78c576/api/src/test/java/org/openmrs/module/patientflags/FlagTest.java#L85), To run these test in success we should have invalid  flags. 

The FlagServiceTest class is mainly targeting flag evaluation process, we should have valid flags (groovy script and sql should work correctly), if there is any invalid logic, the tests may not execute as we expected. These evaluation methods fetch all flags in the database, and if any flag contains invalid logic, the evaluation process will not succeed.

File: api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java (hunk 5 above)
Comment (ManojLL) [reply api/src/test/java/org/openmrs/module/patientflags/FlagServiceTest.java]: I removed the `throws Exception`  from methods

**Pull request merged**